### PR TITLE
Cleanup chain config and unify how to check fork blocks

### PR DESCRIPTION
**Pull request by trinhdn2**
This PR removes ugly fork blocks declared in `common/constants.go` and unifies the way how we check for forking using the methods of `params.ChainConfig`. The mainnet chain config (contains various fork blocks) can be configured in `params/config.go`
```go
// TomoChain mainnet config
TomoMainnetChainConfig = &ChainConfig{
    ChainId:                      big.NewInt(88),
    HomesteadBlock:               big.NewInt(1),
    EIP150Block:                  big.NewInt(2),
    EIP150Hash:                   common.HexToHash("0x0000000000000000000000000000000000000000000000000000000000000000"),
    EIP155Block:                  big.NewInt(3),
    EIP158Block:                  big.NewInt(3),
    ByzantiumBlock:               big.NewInt(4),
    TIP2019Block:                 big.NewInt(1050000),
    TIPSigningBlock:              big.NewInt(3000000),
    TIPRandomizeBlock:            big.NewInt(3464000),
    BlackListHFBlock:             big.NewInt(9349100),
    TIPTomoXBlock:                big.NewInt(20581700),
    TIPTomoXLendingBlock:         big.NewInt(21430200),
    TIPTomoXCancellationFeeBlock: big.NewInt(30915660),
    TIPTRC21FeeBlock:             big.NewInt(13523400),
    Posv: &PosvConfig{
        Period:              2,
        Epoch:               900,
        Reward:              250,
        RewardCheckpoint:    900,
        Gap:                 5,
        FoudationWalletAddr: common.HexToAddress("0x0000000000000000000000000000000000000068"),
    },
    }
```
Similarly, the testnet fork blocks can be hardcoded at the same place
```go
// TestnetChainConfig contains the chain parameters to run a node on the Ropsten test network.
TestnetChainConfig = &ChainConfig{
    ChainId:                      big.NewInt(3),
    HomesteadBlock:               big.NewInt(0),
    DAOForkBlock:                 nil,
    DAOForkSupport:               true,
    EIP150Block:                  big.NewInt(0),
    EIP150Hash:                   common.HexToHash("0x62e0fde86e34c263e250fbcd5ca4598ba8ca10a1d166c8526bb127e10b313311"),
    EIP155Block:                  big.NewInt(10),
    EIP158Block:                  big.NewInt(10),
    ByzantiumBlock:               big.NewInt(1700000),
    ConstantinopleBlock:          nil,
    TIP2019Block:                 big.NewInt(0),
    TIPSigningBlock:              big.NewInt(0),
    TIPRandomizeBlock:            big.NewInt(0),
    BlackListHFBlock:             big.NewInt(0),
    TIPTomoXCancellationFeeBlock: big.NewInt(0),
    TIPTRC21FeeBlock:             big.NewInt(0),
    Ethash:                       new(EthashConfig),
}
```

### CODE REVIEW

File: core/chain_makers.go [hunk 1]
```diff
@@ -108,7 +109,7 @@ func (b *BlockGen) AddTxWithChain(bc *BlockChain, tx *types.Transaction) {
 	b.receipts = append(b.receipts, receipt)
 	if tokenFeeUsed {
 		fee := new(big.Int).SetUint64(gas)
-		if b.header.Number.Cmp(common.TIPTRC21Fee) > 0 {
+		if b.config.IsTIPTRC21Fee(b.header.Number) {
```
Comment (tungng98) [inline core/chain_makers.go]: I doubt that the comparison is incorrect. The hardfork check function use `isForked` internally, which is `s.Cmp(head) <= 0` or head block must be GreaterThanEqual hardfork block. The original statement only have Greater comparison

File: params/config.go [hunk 2]
```diff
@@ -32,13 +32,21 @@ var (
...
+		EIP158Block:                  big.NewInt(3),
+		ByzantiumBlock:               big.NewInt(4),
+		TIP2019Block:                 big.NewInt(1050000),
+		TIPSigningBlock:              big.NewInt(3000000),
+		TIPRandomizeBlock:            big.NewInt(3464000),
+		BlackListHFBlock:             big.NewInt(9349100),
+		TIPTomoXBlock:                big.NewInt(20581700),
+		TIPTomoXLendingBlock:         big.NewInt(21430200),
+		TIPTomoXCancellationFeeBlock: big.NewInt(30915660),
+		TIPTRC21FeeBlock:             big.NewInt(13523400),
```
Comment (tungng98) [inline params/config.go]: According the block number, TIPTRC21FeeBlock should be placed before TIPTomoXBlock

File: params/config.go [hunk 3]
```diff
@@ -49,52 +57,67 @@ var (
...
+		ChainId:        big.NewInt(89),
+		HomesteadBlock: big.NewInt(0),
+		EIP150Block:    big.NewInt(0),
+		EIP155Block:    big.NewInt(0),
+		EIP158Block:    big.NewInt(0),
+		ByzantiumBlock: big.NewInt(0),
+		Posv:           &PosvConfig{Period: 0, Epoch: 30000},
+	}
+
+	TestChainConfig = &ChainConfig{
```
Comment (tungng98) [inline params/config.go:124]: As my understanding, TestChainConfig is used for our testnet. Why all hardfork blocks aren't pre-defined here. Please clarify.

File: core/chain_makers.go (hunk 1 above)
Comment (trinhdn2) [reply core/chain_makers.go]: That's correct, I missed the point. It must be checked with `b.header.Number + 1`

File: params/config.go (hunk 3 above)
Comment (trinhdn2) [reply params/config.go:124]: `TestChainConfig` is used for unit testing only. `TestnetChainConfig` above is the one which is used for the testnet.

Comment (trinhdn2) [reply params/config.go:124]: I'll add comment for this. And we will need to choose the testnet fork blocks before this PR is merged. I suggest the final `TestnetChainConfig` as below
```go
TestnetChainConfig = &ChainConfig{
		ChainId:                      big.NewInt(89),
		HomesteadBlock:               big.NewInt(0),
		DAOForkBlock:                 nil,
		DAOForkSupport:               false,
		EIP150Block:                  big.NewInt(0),
		EIP150Hash:                   common.HexToHash("0x0000000000000000000000000000000000000000000000000000000000000000"),
		EIP155Block:                  big.NewInt(0),
		EIP158Block:                  big.NewInt(0),
		ByzantiumBlock:               big.NewInt(0),
		ConstantinopleBlock:          big.NewInt(0),
		TIP2019Block:                 big.NewInt(0),
		TIPSigningBlock:              big.NewInt(0),
		TIPRandomizeBlock:            big.NewInt(0),
		BlackListHFBlock:             big.NewInt(0),
		TIPTomoXCancellationFeeBlock: big.NewInt(0),
		TIPTRC21FeeBlock:             big.NewInt(0),
		Posv: &PosvConfig{
			Period:              2,
			Epoch:               900,
			Reward:              250,
			RewardCheckpoint:    900,
			Gap:                 5,
			FoudationWalletAddr: common.HexToAddress("0x0000000000000000000000000000000000000068"),
		},
	}
```

Comment (trinhdn2) [reply params/config.go:124]: We also need to define the default bootnodes for the testnet here either
https://github.com/BuildOnViction/tomochain/blob/master/params%2Fbootnodes.go#L41

File: eth/api_tracer.go [hunk 4]
```diff
@@ -701,7 +701,7 @@ func (api *PrivateDebugAPI) computeTxEnv(blockHash common.Hash, txIndex int, ree
 
 		if tokenFeeUsed {
 			fee := new(big.Int).SetUint64(gas)
-			if block.Header().Number.Cmp(common.TIPTRC21Fee) > 0 {
+			if api.config.IsTIPTRC21Fee(block.Header().Number) {
```
Comment (tungng98) [inline eth/api_tracer.go]: Many places changing from block number to ChainConfig.IsXXX still need to reviewed and be fixed.

**Comment by trinhdn2** [comment]
I will break down this PR into smaller ones later

**Pull request closed without merging**